pyro/poutine/__init__.py

```diff
@@ -205,23 +242,26 @@ def queue(fn, queue, max_tries=None,
     if num_samples is None:
         num_samples = -1
 
-    def _fn(*args, **kwargs):
-
-        for i in xrange(max_tries):
-            assert not queue.empty(), \
-                "trying to get() from an empty queue will deadlock"
-
-            next_trace = queue.get()
-            try:
-                ftr = trace(escape(replay(fn, next_trace),
-                                   functools.partial(escape_fn, next_trace)))
-                return ftr(*args, **kwargs)
-            except util.NonlocalExit as site_container:
-                site_container.reset_stack()
-                for tr in extend_fn(ftr.trace.copy(), site_container.site,
-                                    num_samples=num_samples):
-                    queue.put(tr)
-
-        raise ValueError("max tries ({}) exceeded".format(str(max_tries)))
-
-    return _fn
+    def wrapper(wrapped):
+        def _fn(*args, **kwargs):
+
+            for i in xrange(max_tries):
+                assert not queue.empty(), \
+                    "trying to get() from an empty queue will deadlock"
+
+                next_trace = queue.get()
+                try:
+                    ftr = trace(escape(replay(wrapped, trace=next_trace),
+                                       escape_fn=functools.partial(escape_fn,
+                                                                   next_trace)))
+                    return ftr(*args, **kwargs)
+                except util.NonlocalExit as site_container:
+                    site_container.reset_stack()
+                    for tr in extend_fn(ftr.trace.copy(), site_container.site,
+                                        num_samples=num_samples):
+                        queue.put(tr)
+
+            raise ValueError("max tries ({}) exceeded".format(str(max_tries)))
+        return _fn
+
+    return wrapper(fn) if fn is not None else wrapper
```

Comment:
I find the early-return pattern a little less invasive and easier to reason about:
```py
def queue(fn=None, queue=None, ...):
    # support use as a decorator
    if fn is None:
        return lambda fn: queue(fn=fn, queue=queue, ...)
    # ...remainder of queue() unmodified...
```
Do you have a preference either way?

Reply:
It's consistent with the primitive aliases, where it's important to return the `Messenger` object rather than a `lambda` wrapper